notifier/notifier.go

```diff
@@ -221,83 +246,113 @@ func (n *Notifier) collectRogueCallbacks() {
 	}
 }
 
-// Notify posts callback info to j.CallbackURL
-func (n *Notifier) Notify(j *job.Job) error {
-	j.CallbackCount++
+// MonitorDeliveries monitors the delivery reports channel of a backend.
+func (n *Notifier) MonitorDeliveries(ctx context.Context, id string) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case cbInfo, ok := <-n.backends[id].DeliveryReports():
+			if !ok {
+				return
+			}
 
-	err := n.markCbInProgress(j)
-	if err != nil {
-		return err
+			err := n.handleCallbackInfo(cbInfo)
+			if err != nil {
+				n.Log.Printf("Error occured during handling callback info. Operation returned error %s", err)
+			}
+		}
 	}
+}
 
-	cbInfo, err := n.getCallbackInfo(j)
-	if err != nil {
-		return n.markCbFailed(j, err.Error())
-	}
+// handleCallbackInfo handles a callback's delivery result.
+// If the callback has been successfully delivred we increment stats counters
+// and remove the job from storage. Otherwise we mark the callback as failed.
+func (n *Notifier) handleCallbackInfo(cbInfo job.Callback) error {
+	var e string
 
-	cb, err := json.Marshal(cbInfo)
+	j, err := n.Storage.GetJob(cbInfo.JobID)
 	if err != nil {
-		return n.markCbFailed(j, err.Error())
+		e = fmt.Sprintf("\nError: Could not get job %s. Operation returned error: %s", cbInfo.JobID, err)
+		return errors.New(e)
```

Comment:
You can just do `return fmt.Errorf()` instead, so you wouldn't need the `e` variable.